File: 15-3sum/3sum.go
func threeSum(nums []int) [][]int {
	res := [][]int{}
	sort.Ints(nums) // Step 1: Sort array

	for i := 0; i < len(nums)-2; i++ {
		// Skip duplicate first elements
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left, right := i+1, len(nums)-1
		for left < right {
			total := nums[i] + nums[left] + nums[right]
			if total == 0 {
				// Found a valid triplet
				res = append(res, []int{nums[i], nums[left], nums[right]})

				// Skip duplicate values for left
				for left < right && nums[left] == nums[left+1] {
					left++
				}
				// Skip duplicate values for right
				for left < right && nums[right] == nums[right-1] {
					right--
				}
				left++
				right--
			} else if total < 0 {
				left++
			} else {
				right--
			}
		}
	}

	return res
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()  # Step 1: Sort array

        for i in range(len(nums) - 2):
            # Skip duplicate first elements
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total == 0:
                    res.append([nums[i], nums[left], nums[right]])

                    # Skip duplicate values for left
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    # Skip duplicate values for right
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1

        return res
